### PR TITLE
core/namespace: allow using ProtectSubset=pid and ProtectHostname=tru…

### PR DESCRIPTION
**Pull request by brauner**
…e together

If a service requests both ProtectSubset=pid and ProtectHostname=true
then it will currently fail to start. The ProcSubset=pid option
instructs systemd to mount procfs for the service with subset=pid which
hides all entries other than /proc/<pid>. Consequently trying to
interact with the two files /proc/sys/kernel/{hostname,domainname}
covered by ProtectHostname=true will fail.

Fix this by only performing this check when ProtectSubset=pid is not
requested. Essentially ProtectSubset=pid implies/provides
ProtectHostname=true.

### CODE REVIEW

**Comment by brauner** [comment]
See bits of the discussion in https://github.com/systemd/systemd/pull/17957.

File: src/core/namespace.c [hunk 1]
```diff
@@ -2157,7 +2157,10 @@ int setup_namespace(
                                 goto finish;
                 }
 
-                if (ns_info->protect_hostname) {
+                /* If proc is mounted with subset=pid then neither of the two
+                 * paths will exist, i.e. they are implicitly protected by the
+                 * mount option. */
+                if (ns_info->protect_hostname && !ns_info->proc_subset == PROC_SUBSET_PID) {
```
Comment (brauner) [inline src/core/namespace.c]: Oh, one thing I should check is how this is handled if the kernel doesn't support `subset=pid`, i.e. do we at that point already now that `PROC_SUBSET_PID` won't work and will have unset `protect_subset`. If not we need to tweak the patch.

Comment (bluca) [inline src/core/namespace.c]: Maybe use `.ignore = ns_info->proc_subset == PROC_SUBSET_PID` instead? Using proc_subnet_pid can fail (eg, old kernel), in which case this should still apply I think

Comment (bluca) [reply src/core/namespace.c]: see other comment, we gracefully fall back in that case

Comment (brauner) [reply src/core/namespace.c]: Ha, we had the same worry. :)
Changed it to set `.ignore` in case `PROC_SUBSET_PID` is requested.

Comment (brauner) [reply src/core/namespace.c]: fixed

**Review by yuwata — APPROVED** [review]
LGTM. If possible, adding a testcase for the original issue is better.

**Comment by yuwata** [comment]
(BTW, I do not have any good idea about that, neither I think we should support such case, but the ENOENT error for hostname or domainname are also ignored on older kernels which do not support ProtectSubset=.)

**Review by bluca — APPROVED** [review]

**Comment by brauner** [comment]
> (BTW, I do not have any good idea about that, neither I think we should support such case, but the ENOENT error for hostname or domainname are also ignored on older kernels which do not support ProtectSubset=.)

I'm not sure what you mean by "neither I think we should support such case". Are you saying we shouldn't support `ProcSubset=pid` and `ProtectHostname=true` used together? I find that to be an unlikely reading of what you're trying to say. :)

It seems you worry about the case where your `subset=pid` isn't supported and you still get `ENOENT` for `/proc/sys/kernel/{domainname,hostname}`? That seems impossible and so not worth worrying imho.

**Comment by brauner** [comment]
Also, are there released systemd versions that ignore `ENOENT` from `{domainname,hostname}` that would hide this issue before the revert?

**Comment by bluca** [comment]
> Also, are there released systemd versions that ignore `ENOENT` from `{domainname,hostname}` that would hide this issue before the revert?

No, I don't think so

**Comment by yuwata** [comment]
> > (BTW, I do not have any good idea about that, neither I think we should support such case, but the ENOENT error for hostname or domainname are also ignored on older kernels which do not support ProtectSubset=.)
> 
> I'm not sure what you mean by "neither I think we should support such case". Are you saying we shouldn't support `ProcSubset=pid` and `ProtectHostname=true` used together? I find that to be an unlikely reading of what you're trying to say. :)
> 
> It seems you worry about the case where your `subset=pid` isn't supported and you still get `ENOENT` for `/proc/sys/kernel/{domainname,hostname}`? That seems impossible and so not worth worrying imho.

Sorry for my unclear comment.

I mean, previously, error in mounting domainname or hostname are always handled as critical. But with this change, if `ns_info->proc_subset == PROC_SUBSET_PID`, then an occasional and accidental error in mounting one of them with ENOENT is ignored even for older kernels witch do not support the new procfs mount option.
I do not have any good way to make such error critical for older kernels. But if there exists a simple way to handle them for older kernels, then that's better.

Is it clearer now?

Anyway, I think it is not necessary to be handled in this PR. It's a kind of off-topic here.

**Comment by brauner** [comment]
> > > (BTW, I do not have any good idea about that, neither I think we should support such case, but the ENOENT error for hostname or domainname are also ignored on older kernels which do not support ProtectSubset=.)
> > 
> > 
> > I'm not sure what you mean by "neither I think we should support such case". Are you saying we shouldn't support `ProcSubset=pid` and `ProtectHostname=true` used together? I find that to be an unlikely reading of what you're trying to say. :)
> > It seems you worry about the case where your `subset=pid` isn't supported and you still get `ENOENT` for `/proc/sys/kernel/{domainname,hostname}`? That seems impossible and so not worth worrying imho.
> 
> Sorry for my unclear comment.
> 
> I mean, previously, error in mounting domainname or hostname are always handled as critical. But with this change, if `ns_info->proc_subset == PROC_SUBSET_PID`, then an occasional and accidental error in mounting one of them with ENOENT is ignored even for older kernels witch do not support the new procfs mount option. I do not have any good way to make such error critical for older kernels. But if there exists a simple way to handle them for older kernels, then that's better.
> 
> Is it clearer now?

Yes, thank you!

It shouldn't be super hard to create a dependency between the procfs mount and these two mounts. Once you have that you can e.g. add a callback function `.ingnorable()` that you can call. Of course, you can only create dependencies on mounts that happened earlier.

**Comment by brauner** [comment]
> > > > (BTW, I do not have any good idea about that, neither I think we should support such case, but the ENOENT error for hostname or domainname are also ignored on older kernels which do not support ProtectSubset=.)
> > > 
> > > 
> > > I'm not sure what you mean by "neither I think we should support such case". Are you saying we shouldn't support `ProcSubset=pid` and `ProtectHostname=true` used together? I find that to be an unlikely reading of what you're trying to say. :)
> > > It seems you worry about the case where your `subset=pid` isn't supported and you still get `ENOENT` for `/proc/sys/kernel/{domainname,hostname}`? That seems impossible and so not worth worrying imho.
> > 
> > 
> > Sorry for my unclear comment.
> > I mean, previously, error in mounting domainname or hostname are always handled as critical. But with this change, if `ns_info->proc_subset == PROC_SUBSET_PID`, then an occasional and accidental error in mounting one of them with ENOENT is ignored even for older kernels witch do not support the new procfs mount option. I do not have any good way to make such error critical for older kernels. But if there exists a simple way to handle them for older kernels, then that's better.
> > Is it clearer now?
> 
> Yes, thank you!
> 
> It shouldn't be super hard to create a dependency between the procfs mount and these two mounts. Once you have that you can e.g. add a callback function `.ingnorable()` that you can call. Of course, you can only create dependencies on mounts that happened earlier.

There are other methods of course. :)

**Comment by yuwata** [comment]
@brauner Thank you for the suggestion. It maybe better to handled with ENOENT handling issue in ExtensionImage/ExtensionDIrectory=.

(Feel free to upgrade the green label if you are planning to add a testcase in later PR.)

**Review by topimiettinen — CHANGES_REQUESTED** [review]
Could you reuse `ignore_protect_proc` above? It would also consider `ns_info->ignore_protect_paths`.

**Pull request merged**